### DISCUSSION TITLE
Meshctrl.js has no output

### DISCUSSION THREAD

**Discussion by timaw7** [Ylianst/MeshCentral#5882] [General]
**Describe the bug**
Running the meshctrl.js file is producing no output not indicating whether it is working or not

**To Reproduce**
1. Login to the server and locate the node_module files
2. Run `node node_modules/meshcentral/meshctrl.js serverinfo`
3. This is also the same for  `node node_modules/meshcentral/meshctrl.js listusers`

**Expected behavior**
It is expected that the outputted results would be as expected or some error to indicate a failed server connection.

**Server Software (please complete the following information):**
 - OS: CentOS
 - Virtualization: None
 - Version:1.1
 - Node: 14.21.1

**Additional context**
My intention is to verify this method before moving onto a remote login.  My understanding from the youtube video is no user credentials are required here.  I'm tried shutting off the server which does produce the expected result.

> Unable to connect to wss://localhost/control.ashx

I've tried the method stated here as well, booting the server manually and but to no avail.

https://github.com/Ylianst/MeshCentral/issues/5198#issuecomment-1596249352

**config.json file**
```
{
  "$schema": "http://info.meshcentral.com/downloads/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "cert": "asdf",
    "WANonly": true,
    "_LANonly": true,
    "_sessionKey": "asdf",
    "_port": 443,
    "_aliasPort": 443,
    "_redirPort": 80,
    "_redirAliasPort": 80,
    "agentPort": 444,
    "agentAliasPort": 443,
    "agentAliasDNS": "asdf",
    "maxInvalidLogin": {
      "time":100,
      "count": 100,
      "coolofftime": 10
    },
    "authLog": "/opt/meshcentral/LoginAttemptsLog"
  },
  "domains": {
    "": {
      "loginKey": [],
      "title": "",
      "_title2": "Servername",
      "welcomePicture": "A_black_background.png",
      "welcomePictureFullScreen": true,
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true,
      "passwordRequirements": {
        "min": 12,
        "max": 50,
        "upper": 1,
        "lower": 1,
        "numeric": 1,
        "nonalpha": 1,
        "reset": 60,
        "force2factor": true,
        "email2factor": false,
        "sms2factor": false,
        "oldPasswordBan": 20,
        "loginTokens": false
      },
      "twoFactorCookieDurationDays": 0
    }
  },
  "letsencrypt": {
  }
 }
```

**Comment by si458** [comment]
this is because you need to provide `meshctrl.js` some login information to talk to your meshcentral server
the file is designed so you can copy it to another machine and use it from there too...
`node meshctrl.js ListEvents --url wss://mymeshserver.com --loginuser userhere --loginpass passhere` this is the basic config
you can also run `node meshctrl.js` and it will list all the options including the login variables at the bottom

**Comment by timaw7** [comment]
I have confirmed I am able to connect when providing the correct credentials on the server and remotely.  Originally, I was following the video guide here and made some assumptions.  I was ssh'd into the server and presumed in this format credentials were not needed.  The script currently provides no output that it succeeded or failed until some authentication credentials are provided.

https://www.youtube.com/watch?v=p_jdtYJyuBQ

Also I'm not sure if this is the right place, but I am curious about the scripting now and if there is some sort of generated token method?  I can go ahead and use my user account or setup a new account for the scripting purpose.  The main nuisance is 2FA.  I'd rather have a token with an expiration than a separate admin account with 2FA shut off.

**Reply by si458** [reply]
you can send a 2fa token with the meshctrl.js `node meshctrl.js`
```
Supported login arguments:
  --url [wss://server]        - Server url, wss://localhost:443 is default.
                              - Use wss://localhost:443?key=xxx if login key is required.
  --loginuser [username]      - Login username, admin is default.
  --loginpass [password]      - Login password OR Leave blank to enter password at prompt
  --token [number]            - 2nd factor authentication token.
  --loginkey [hex]            - Server login key in hex.
  --loginkeyfile [file]       - File containing server login key in hex.
  --logindomain [domainid]    - Domain id, default is empty, only used with loginkey.
  --proxy [http://proxy:123]  - Specify an HTTP proxy.
```

you can also create a Login Token to use with meshctrl.js instead from the web ui 
click 'My Account' then click create 'Login Token'
and this created unique user/pass to be used instead

**Reply by timaw7** [reply]
I am.  It is just a matter of convenience.  Essentially migrating the scripting to a scripting account as opposed to using a personal account with standard login credentials.

**Reply by si458** [reply]
In theory u can use meshctrl with ur user login details to then create users with there own login deyails/tokens, best run `node meshctrl.js` to get all the options then run `node meshctrl.js help METHODHERE` to get more info about that method
